### ISSUE TITLE
Mouse wheel lockup

### ISSUE THREAD

**Issue by ovlx** [pekwm/pekwm#4]
I have encountered an issue where, if you scroll really up and down using the mouse wheel, all mouse inputs will lockup randomly, no left or right click, pressing space(and otherkeys) on the keyboard interrupts the lock and it continue as normal 

running 18.04.3 with lxqt  and kernels 5.0/5.2/5.3 
disabling any keybinds to the mousewheel had no effect.
setting GDK_CORE_DEVICE_EVENTS=1 had no effect.

mouse is set to 500hz, will update on whether dropping the polling rate to 125hz changes anything.
edit: noeffect
latest git version was compilled using -O2

**Comment by biopsin** [comment]
This gets triggered when scrolling up/down with mouse wheel in a window now and then, I noticed that it waits on event and pressing mod4 it unblocks and in my case
mod4+mouse(up/down) moves window to next desktop, so it suddently jumps to next desktop...odd

**Comment by biopsin** [comment]
Last comment on this issue; ..and it only triggers for me on a display resolution above 1920x1080 with same config files and so on (tested on two separate pc 4K vs HD). 
I've being reading the source, but I'm to dumb to figure out the connection.

OFF TOPIC : Geez can't even edit my post anymore as Github is breaking any other webengine with their Chrome only webcomponents regime; time to piss off again..

**Comment by pigge2** [comment]
i also experiences this, and it appears to have gotten worse since my last system update (which did not include pekwm 0.18, but i do run 0.18 compiled from git since 2020-11-24) and i cannot recreate the error on MATE. 
Some times xorg crashes completely when i "unfreeze" the windows, and at those times gdb tells me a SIGSEGV was received at Observable.cc:27 (but i'm not sure these two errors really are related)

**Comment by biopsin** [comment]
@pekdon while testing out branch zombie-14, any recommend tool I should use to track this issue as well?

**Comment by pekdon** [comment]
To be honest I'm not sure. xev might provide relevant information but it will be for a specific Window and will also produce so much output it will be hard to figure out what is relevant.

Consistent reproduction steps would help but that is easier said than done. Haven't experienced it myself in a long while.

**Comment by biopsin** [comment]
~ $ xev -root

Focus and raise window (mouse leftclick)
```
PropertyNotify event, serial 18, synthetic NO, window 0x6de,
    atom 0x1 (PRIMARY), time 17915423, state PropertyNewValue

ConfigureNotify event, serial 18, synthetic NO, window 0x6de,
    event 0x6de, window 0xc002e9, (640,10), width 2560, height 2140,
    border_width 0, above 0xc002c2, override YES

PropertyNotify event, serial 18, synthetic NO, window 0x6de,
    atom 0x150 (_NET_CLIENT_LIST_STACKING), time 17915423, state PropertyNewValue

PropertyNotify event, serial 18, synthetic NO, window 0x6de,
    atom 0x156 (_NET_ACTIVE_WINDOW), time 17915424, state PropertyNewValue

```

Scrollwheel up

```
PropertyNotify event, serial 21, synthetic NO, window 0x6de,
    atom 0x1 (PRIMARY), time 17928415, state PropertyNewValue

PropertyNotify event, serial 21, synthetic NO, window 0x6de,
    atom 0x150 (_NET_CLIENT_LIST_STACKING), time 17928415, state PropertyNewValue
```

Scrollwheel down

```
PropertyNotify event, serial 21, synthetic NO, window 0x6de,
    atom 0x1 (PRIMARY), time 17930066, state PropertyNewValue

PropertyNotify event, serial 21, synthetic NO, window 0x6de,
    atom 0x150 (_NET_CLIENT_LIST_STACKING), time 17930066, state PropertyNewValue
```

but while it locks up so does xev output and a
```
FocusIn event, serial 21, synthetic NO, window 0x6de,
    mode NotifyGrab, detail NotifyInferior
```
brakes lockup

**Comment by pekdon** [comment]
@biopsin it would be real interesting to see what happens if you clear out all but the bare minimum of key and mouse bindings, especially the ones on button 4/5 and see if you still can reproduce the issue!

I think I will be adding some trace logging on grab/ungrab operations as well to further be able to understand the issue.

**Comment by biopsin** [comment]
@pekdon; Your right I do have set mouse button bindings on 4/5 in Client  `Actions = "Focus; Raise"` 
was set to try to compensate the lockups and then got used to it.. :oops
I'll test again a complete default setup with a bare minimum bindings for keys&mouse

**Comment by biopsin** [comment]
Testing latest master with --log-file issue.log --log-level trace ; but it's unfortunatly not catched here either..
([log@teknik](https://p.teknik.io/Simple/xz0QM)) - Link expires in 2 days
However it's getting much more difficult to trigger it now then on v0.1.8

**Comment by pigge2** [comment]
i did a rebuild of master, and tried trace -logging as well, and i can't see anything either in the trace-log when the error occurs, but i can provoke the error in a few secconds if i scroll quickly up and down with my trackpad in slack and Vivaldi, but also some times in RDP sessions (where it appears as if the shift-key is stuck, until i "release" the mouse pointer with a single press on the win-key) i think it might be in Electron apps most notably...

**Comment by ovlx** [comment]
I have also discovered that the bug can also  be  triggered when just rapidly clicking ui elements (the window decorations for instance)  it seems , so its not specifically the mouse wheel.

**Comment by pekdon** [comment]
Seems as if I'll have to harder to reproduce this locally, thanks for all the information so hard. Won't have time to look into until earliest next week.

**Comment by maoamid** [comment]
I had the same issue with initial v0.1.18, but after I compiled the last revision (v0.2.0) the problem doesn't showed up anymore using the same configs... maybe someone can confirm

**Comment by pekdon** [comment]
What is latest? (Githash)

**Comment by biopsin** [comment]
Well I run @ [003cab5697ee205a35ef3124b7ab960b70b3dbf3] and I still have lockups just less frequantly, however I have notice lockups as ovlx mentioned too, something that has not been so obvious as I use theme without decorations.

**Comment by pekdon** [comment]
@biopsin mind trying d7616d3 out, had limited amount of time testing it out but might have improved the situation.

**Comment by biopsin** [comment]
@pekdon I trided building it but it fails

> ../src/x11.hh: In static member function 'static void X11::destroyImage(XImage*)':
> ../src/x11.hh:773:13: error: 'XDestroyImage' was not declared in this scope; did you mean 'destroyImage'?
>   773 |             XDestroyImage(ximage);
>       |             ^~~~~~~~~~~~~
>       |             destroyImage
> [9/65] Building CXX object src/CMakeFiles/x11.dir/PWinObjReference.cc.o
> FAILED: src/CMakeFiles/x11.dir/PWinObjReference.cc.o
> /usr/bin/g++ -DBINDIR=\"/usr/bin\" -DDATADIR=\"/usr/share\" -DFEATURES="\" Xft\"" -DSYSCONFDIR=\"/etc/pekwm\" -DVERSION=\"0.2.0\" -Isrc -I/usr/include/freetype2 -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -std=c++11 -O3 -DNDEBUG -MD -MT src/CMakeFiles/x11.dir/PWinObjReference.cc.o -MF src/CMakeFiles/x11.dir/PWinObjReference.cc.o.d -o src/CMakeFiles/x11.dir/PWinObjReference.cc.o -c ../src/PWinObjReference.cc
> In file included from ../src/PWinObj.hh:17,
>                  from ../src/PWinObjReference.hh:13,
>                  from ../src/PWinObjReference.cc:11:
> ../src/x11.hh: In static member function 'static void X11::destroyImage(XImage*)':
> ../src/x11.hh:773:13: error: 'XDestroyImage' was not declared in this scope; did you mean 'destroyImage'?
>   773 |             XDestroyImage(ximage);
>       |             ^~~~~~~~~~~~~
>       |             destroyImage
> [10/65] Building CXX object src/CMakeFiles/pekwm_panel.dir/pekwm_panel.cc.o
> [11/65] Building CXX object src/CMakeFiles/util.dir/Util.cc.o
> ninja: build stopped: subcommand failed.

so I asumed things are not finnished yet.

**Comment by pekdon** [comment]
Oh, what are you building on? Verified OK on Ubuntu, NetBSD 9.1, FreeBSD 12 and OpenBSD 6.8

**Comment by biopsin** [comment]
voidlinux build container

**Comment by biopsin** [comment]
ahh, I might missing a dependcy..

**Comment by pekdon** [comment]
> voidlinux build container

pull and try again (should add that to my buildbot setup as well)

**Comment by biopsin** [comment]
nope; same error.. all options left to default

> configure_args="-DCMAKE_INSTALL_SYSCONFDIR=/etc"
> makedepends="libjpeg-turbo-devel libpng-devel libXinerama-devel
>  libXft-devel libXrender-devel fontconfig-devel libXpm-devel libXrandr-devel"

**Comment by biopsin** [comment]
[4f08bc9a8cb6cd4a0a52c8e91ebd849d201e3ae2] build fine again ,tnx!

**Comment by pekdon** [comment]
Do you have any mouse client window mouse bindings for button 4 or 5? Still have set GDK_CORE_DEVICE_EVENTS to 1? If so, try setting to 0 and remove those bindings and see if it still happens.

**Comment by pigge2** [comment]
i cannot re-produce any lockups either.. i do have some weirdness going on with my mouse, which almost feels like interrupts/dma something, but that might be entirely unrelated to pekwm, i haven't had the time to investigate properly yes

**Comment by biopsin** [comment]
Sorry @pekdon it does not make any difference. 
Tried: removed keybinds for 4&5, flipping GDK.. 0 then 1, I restarted and reloaded inbetween flips.
@pigge2 can you try a site with many [pages](https://wiki.archlinux.org/index.php/GTK), scroll fast down and up couple of times.

**Comment by pigge2** [comment]
> Sorry @pekdon it does not make any difference.
> Tried: removed keybinds for 4&5, flipping GDK.. 0 then 1, I restarted and reloaded pekwm inbetween flips.
> @pigge2 can you try a site with many [pages](https://wiki.archlinux.org/index.php/GTK), scroll fast down and up couple of times.

no, i can't get it to lock up no matter what i do to abuse it, i previously could get it to lockup in less than 5 seconds without exception in Slack, now i've scrolled back 3 months in a #general channel with lots of chatter, and tried the archwiki page, all it did was to make my machine hot, but no lock-ups

**Comment by biopsin** [comment]
@pigge2 thank you for testing.

**Comment by biopsin** [comment]
thing is it happening in any client window, tested now scrolling back in simple terminal and it ain't even using gtk.

**Comment by pigge2** [comment]
> thing is it happening in any client window, tested now scrolling back in simple terminal and it ain't even using gtk.
> Edit: also tested disabling compositor (xcompmgr)

yeah, i just had it happen again.. this time i was pasting text into sublime edit, and the window locked up..

**Comment by pekdon** [comment]
So for you @biopsin it got worse by the latest change, and this being on void linux I guess? (Probably time to get void up and running on my Pi then)

**Comment by maoamid** [comment]
> What is latest? (Githash)

Yes, I build it from git yesterday (d7616d3) (with all deps installed, cmake, make, make install ... in Void Linux on a Toshiba laptop, I had some troubles running make with an error "XDestroyImage not defined" but I fixed it installing some devel libs and xwininfo )

Let me explain more, maybe helps (today I just had it happen again with or without compton / picom no matter to me):
It happens if I use a mouse (using touchpad I can't reproduce it yet)

- How I reproduce it? copy some text from Firefox, paste it somewhere (VSCode), press alt+tab a few times, ctrl+c / ctrl+v / ctr+z after (it lock up most of the time doing this)
- After a lock up if I push any key all is back normally
- If after a lock up I go to another view ("Super + n") and back to current view, the lock up is also gone but the window that "caused" the lock up is moved/positioned at other coordinates (very strange, it happens most of the time)

**Comment by biopsin** [comment]
> So for you @biopsin it got worse by the latest change, 
No, as mentioned earlier its less frequently, but still happens.
It almost feels like some buffer filling up and then need a purge by a keypress, so it all depends on how much scrolling I do. (..makes no sence :D )

**Comment by pekdon** [comment]
It kind of feels like these actions are being triggered:

```
        ButtonRelease = "Mod1 4" { Actions = "SendToWorkspace Next; GotoWorkspace Next" }
        ButtonRelease = "Mod1 5" { Actions = "SendToWorkspace Prev; GotoWorkspace Prev" }
        ButtonRelease = "Mod1 Shift 4" { Actions = "SendToWorkspace PrevV; GotoWorkspace PrevV" }
        ButtonRelease = "Mod1 Shift 5" { Actions = "SendToWorkspace NextV; GotoWorkspace NextV" }
```

I will add some tracing of actions executed in the Action Handler to see if that sheds more light on the issue.

**Comment by biopsin** [comment]
Well tried disabling my bindings
```
Client {
	#Enter = "Any Any" { Actions = "Focus" }
	ButtonPress = "1" { Actions = "HideAllMenus; Focus; Raise" }
	ButtonPress = "3" { Actions = "HideAllMenus" }
	ButtonPress = "4" { Actions = "Focus; Raise" }
	ButtonPress = "5" { Actions = "Focus; Raise" }
	ButtonRelease = "Mod4 1" { Actions = "Lower; UnFocus" }
	ButtonRelease = "Mod4 2" { Actions = "Exec $LOCAL/skippy" }
	ButtonRelease = "Mod4 3" { Actions = "HideAllMenus; ShowMenu Window" }
	#ButtonPress = "Mod4 4" { Actions = "SendToWorkspace PrevV; Exec $DESKID; GotoWorkspace PrevV" }
	#ButtonPress = "Mod4 5" { Actions = "SendToWorkspace NextV; Exec $DESKID; GotoWorkspace NextV" }
	Motion = "Mod4 1" { Threshold = "5"; Actions = "Move" }
	Motion = "Mod4 3" { Threshold = "5"; Actions = "Resize" }
}

Root {
	ButtonPress = "1" { Actions = "HideAllMenus" }
	ButtonRelease = "2" { Actions = "Exec $LOCAL/skippy" }
	ButtonRelease = "3" { Actions = "HideAllMenus; ShowMenu Root" }
	#ButtonPress = "4" { Actions = "Exec $DESKID; GoToWorkspace PrevV" }
	#ButtonPress = "5" { Actions = "Exec $DESKID; GoToWorkspace NextV" }
}
```
but made no difference, unless they get silentry triggered but without actual action preformed.

**Comment by pekdon** [comment]
Do you include the system configuration (as they have them) or you have all own bindings @biopsin?

**Comment by biopsin** [comment]
Se commen above

**Comment by biopsin** [comment]
I can finally confirm
```
	#ButtonPress = "4" { Actions = "Focus; Raise" }
	#ButtonPress = "5" { Actions = "Focus; Raise" }
```
in combination with
```
	#ButtonPress = "Mod4 4" { Actions = "SendToWorkspace PrevV; Exec $DESKID; GotoWorkspace PrevV" }
	#ButtonPress = "Mod4 5" { Actions = "SendToWorkspace NextV; Exec $DESKID; GotoWorkspace NextV" }
```
on Client is the culprint

**Comment by pekdon** [comment]
@ovlx tried current git master with an updated mouse config to see if fixes the issues for you?

**Comment by ovlx** [comment]
@pekdon seems to be fixed, thank you for your persistence!

**Comment by pekdon** [comment]
Thanks for testing @ovlx, at least I know where to look now.

**Comment by biopsin** [comment]
Had to eventually remove those last ButtonPress 4/5 as well, had sporadically but very very seldom lockups.
After now two weeks of testing I no longer experience it.

**Comment by stuart-mclaren** [comment]
In case it helps, I'm seeing this too on FreeBSD 13.0: pekwm-0.1.17_5,1
Using the mouse scroll wheel on Chromium reproduces fairly frequently.

**Comment by biopsin** [comment]
@stuart-mclaren is this a default install with no custom configuration?

**Comment by biopsin** [comment]
After some more frantic scrolltestes on different long sites past days, Im happy to announce this bug is no longer happening for me building against master, so the last refactor of events have a positive effect. Great work!

**Comment by pekdon** [comment]
Unfortunately I'm still able to produce lockups using gvim clicking franticly between the two windows as explained by gammy.

Also, the current master have had some issues for me on Debian 11 amd64 but seems OK on NetBSD 9.2.

Mind trying out the button-events branch and see how that behaves for you @biopsin ?

**Comment by biopsin** [comment]
@pekdon but it does not trigger on the wheel so I salute it 👍🏻

**Comment by maoamid** [comment]
🤔 maybe I'm wrong but for me it looks like after a few clicks when the window manager seems frozen, it is behaving very similar when like a key press is required in a key chain sequence

**Comment by gammy** [comment]
I concur with @pekdon and @maoamid, my "frantic click" issue persists. I just got back from a holiday and haven't yet had time to "frantically scroll" though. Excited to see some progress regardless!

**Comment by biopsin** [comment]
@pekdon & button-events branch : there is something happening for me with the mouse -> Mod4 + [button] where nothing get triggered. Its bit weird as it did work yesterday when testing hmm. I have to switch to debug mode and try again..

**Comment by AstoriaFloyd** [comment]
After the latest commit with the extra flush removed I have been unable to get pekwm to lock up.

**Comment by biopsin** [comment]
I still experience one odd issue; sometimes when selecting on any applications top menubar the dropdownmenu wont expand or its hidden behind the main window, either way a Pekwm restart resolves it.
But else I concure with @AstoriaFloyd

**Comment by pekdon** [comment]
Closing this issue, haven't had the lockups myself. If you're able to reproduce the menubar issue @biopsin, do open a new ticket with notes on how to reproduce it!

**Comment by nwrkbiz** [comment]
FYI: having the same issue. For me it happens when scrolling fast. It is easy to reproduce when using chrome. To unblock the WM again I either need to hit the escape button a few times, or switch back and forth between two TTYs using "ctrl + alt + any F key".

I am using the debian stable stock version of pekwm.

**Comment by biopsin** [comment]
@nwrkbiz 

> I am using the debian stable stock version of pekwm.
what output from `pekwm --version` ?
I guess its solved in master so you will have to build it for the time being.

**Comment by nwrkbiz** [comment]
> @nwrkbiz
> 
> > I am using the debian stable stock version of pekwm.
> 
> what output from `pekwm --version` ?
> 
> I guess its solved in master so you will have to build it for the time being.

pekwm: version 0.1.18

**Comment by AstoriaFloyd** [comment]
The debian repo version of pekwm is nearly two years outdated, and you should update to the latest release.
